src/Workspaces/Core/Portable/Options/SerializableOptionSet.cs

```diff
@@ -17,66 +17,64 @@
+        private ImmutableDictionary<OptionKey, object?> _values;
+
+        private readonly ImmutableHashSet<OptionKey> _changedOptionKeys;
 
         private SerializableOptionSet(
-            WorkspaceOptionSet workspaceOptionSet,
-            ImmutableHashSet<OptionKey> changedOptionKeysNonSerializable)
+            IOptionService globalOptions,
+            ImmutableDictionary<OptionKey, object?> values,
+            ImmutableHashSet<OptionKey> changedOptionKeys)
         {
-            _workspaceOptionSet = workspaceOptionSet;
-            _changedOptionKeysNonSerializable = changedOptionKeysNonSerializable;
+            _globalOptions = globalOptions;
+            _values = values;
+            _changedOptionKeys = changedOptionKeys;
         }
 
-        internal SerializableOptionSet(
-            IOptionService optionService)
-            : this(new WorkspaceOptionSet(optionService), changedOptionKeysNonSerializable: ImmutableHashSet<OptionKey>.Empty)
+        internal SerializableOptionSet(IOptionService globalOptions)
+            : this(globalOptions, values: ImmutableDictionary<OptionKey, object?>.Empty, changedOptionKeys: ImmutableHashSet<OptionKey>.Empty)
         {
         }
 
         [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/30819", AllowLocks = false)]
         private protected override object? GetOptionCore(OptionKey optionKey)
         {
-            return _workspaceOptionSet.GetOption(optionKey);
+            if (_values.TryGetValue(optionKey, out var value))
+            {
+                return value is ICodeStyleOption codeStyleOption ? codeStyleOption.AsPublicCodeStyleOption() : value;
+            }
+
+            value = _globalOptions.GetOption(optionKey);
+            return ImmutableInterlocked.GetOrAdd(ref _values, optionKey, value);
         }
 
         public override OptionSet WithChangedOption(OptionKey optionKey, object? value)
         {
             // Make sure we first load this in current optionset
-            var currentValue = this.GetOption(optionKey);
+            var currentValue = GetOption(optionKey);
 
             // Check if the new value is the same as the current value.
             if (Equals(value, currentValue))
             {
                 // Return a cloned option set as the public API 'WithChangedOption' guarantees a new option set is returned.
-                return new SerializableOptionSet(
-                    _workspaceOptionSet, _changedOptionKeysNonSerializable);
+                return new SerializableOptionSet(_globalOptions, _values, _changedOptionKeys);
             }
 
             return new SerializableOptionSet(
-                (WorkspaceOptionSet)_workspaceOptionSet.WithChangedOption(optionKey, value),
-                _changedOptionKeysNonSerializable.Add(optionKey));
+                _globalOptions,
+                _values.SetItem(optionKey, value),
+                _changedOptionKeys.Add(optionKey));
```

Comment:
The uses of `_values` and `_changedOptionKeys` was a little confusing to me at first.  Am wondering if it would make sense to store a single map of changedKey:changedValue (assuming of course that the `_globalOptions.GetOption` doesn't have perf concerns).

Reply:
Yeah, so I was being careful here. `_globalOptions.GetOption` is locking and reading/writing option value to persisters. I think the reason why we have separate `_changedOptionKeys` is that the _`values` serve as a cache for options that are read from the global options but not changed.

Reply:
also a bit unclear to me.  consider doc'ing more.